utils/spdx/src/main/kotlin/parser/SpdxExpressionLexer.kt

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (C) 2024 The ORT Project Authors (see <https://github.com/oss-review-toolkit/ort/blob/main/NOTICE>)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.utils.spdx.parser
+
+import org.ossreviewtoolkit.utils.common.nextOrNull
+
+/**
+ * A lexer for SPDX expressions. It consumes a sequence of characters and produces a sequence of [Token]s. For details
+ * on the grammar see [SpdxExpressionParser].
+ */
+class SpdxExpressionLexer(input: Sequence<Char>) {
+    constructor(input: String) : this(input.asSequence())
+
+    companion object {
+        private val UPPERCASE = 'A'..'Z'
+        private val LOWERCASE = 'a'..'z'
+        private val DIGITS = '0'..'9'
+
+        private fun Char.isUpper() = this in UPPERCASE
+        private fun Char.isLower() = this in LOWERCASE
+        private fun Char.isDigit() = this in DIGITS
```

Comment:
Why not use the Kotlin built-in functions?

Reply:
The built-in functions would allow many UTF characters that are not allowed by the SPDX specs, I have added docs.

Reply:
Yes, the specs indeed just call it `ALPHA` in the grammar. I assumed that meant the standard English alphanumeric characters, but it could also many any alphanumeric characters. Anyway, the previous implementation supported the same character set, so I think we could discuss this also after merge.